docs/rfcs/rfc-22-projection-subquery.md

```diff
@@ -0,0 +1,180 @@
+  movies(where: {released: 1999}) {
+    title
+    actors(where: {name: "Keanu Reeves"}) {
+      name
+    }
+  }
+}
+```
+
+```cypher
+MATCH (this:`Movie`)
+WHERE this.released = $param0
+RETURN this { .title, actors: [ (this)<-[:ACTED_IN]-(this_actors:Person)  WHERE this_actors.name = $this_actors_param0 | this_actors { .name } ] } as this
+```
+
+This means that top-level projection and filtering is inherently different to nested filtering, and the former happens as part of a `MATCH` or `CREATE` clause and the 
+later is limited to a comprehension. This causes code duplicity, as the same behaviour needs to be implemented in both places, and limits 
+adding features and performance improvements in nested fields. For instance aggregation require `runFirstColumn` to workaround this limitation and sort uses `apoc.coll.sortMulti`:
+
+```graphql
+query Query {
+  movies {
+    actorsAggregate {
+      node {
+        name {
+          longest
+        }
+      }
+    }
+  }
+}
+```
+
+
+```cypher
+MATCH (this:`Movie`)
+RETURN this { actorsAggregate: { node: { name: head(apoc.cypher.runFirstColumnMany("MATCH (this)<-[r:ACTED_IN]-(n:Person)    
+        WITH n as n
+        ORDER BY size(n.name) DESC
+        WITH collect(n.name) as list
+        RETURN {longest: head(list), shortest: last(list)}", { this: this })) } } } as this
+```
+
+## Proposed Solution
+This RFC proposes to move the projection out of the return clause.
+
+Using `OPTIONAL MATCH` we can perform multiple matches and collect them. A list comprehension can be used to return them in the same way as before:
+```cypher
+MATCH (this:`Movie`)
+WHERE this.released = 1999
+OPTIONAL MATCH (this)<-[:ACTED_IN]-(this_actors:Person)
+WHERE this_actors.name = "Keanu Reeves"
+WITH this, collect(this_actors) as this_actors
+RETURN this { .title, actors: [x IN this_actors | x { .name }]}
+```
+
+
+Using `CALL`, both the `OPTIONAL` and `WITH` statements are not needed, leaving a slightly easier to compose query:
+```cypher
+MATCH (this:`Movie`)
+WHERE this.released = 1999
+CALL {
+    WITH this
+    MATCH (this)<-[:ACTED_IN]-(this_actors:Person)
+    WHERE this_actors.name = "Keanu Reeves"
+    RETURN collect(this_actors {.name}) as this_actors
+}
+RETURN this { .title, actors: this_actors }
+```
+
+This option, however, may have negative performance implications, over the current PatternComprehension and `OPTIONAL MATCH`.
+
+By using CALL, it would be easier to compose in a nested fashion:
+
+```cypher
+MATCH (this:`Movie`)
+WHERE this.released = 1999
+CALL {
+    WITH this
+    MATCH (this)<-[:ACTED_IN]-(this_actors:Person)
+    WHERE this_actors.name = "Keanu Reeves"
+    CALL {
+      WITH this_actors
+      MATCH(this_actors)-[:DIRECTED]->(this_movies:Movie)
+      WHERE m.released > 1999
+      RETURN collect(this_movies {.title}) as this_movies
+    }
+    RETURN collect(this_actors {.name, directed: this_movies}) as this_actors
+}
+RETURN this { .title, actors: this_actors}
+```
+
```

Comment:
Which alternative are we using now? `OPTIONAL MATCH` or `CALL` subqueries? Will it be a mix?

Reply:
I just updated the RFC, we will go with `CALL` as it is the one already used in some places, as well as the easier one.
I will keep the optional match as an alternative solution, as it may lead to performance improvements in the future